### PR TITLE
[1.3.9-prepare]Add standalone server module to make it easier to develop and run #6022

### PR DESCRIPTION
**Pull request by CalvinKirs**

### CODE REVIEW

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java [hunk 1]
```diff
@@ -56,6 +58,9 @@
     private int maxWaitTime;
 
     public String getServerList() {
+        if(StringUtils.isNotBlank(System.getProperty("zookeeper.quorum"))){
+            return System.getProperty("zookeeper.quorum");
+        }
```
Comment (CalvinKirs) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java]: Avoid modifying too much, do this temporarily and the dev version is OK.

**Comment by CalvinKirs** [comment]
@kezhenxu94 hi, can you help me review?thx

**Comment by kezhenxu94** [comment]
> @kezhenxu94 hi, can you help me review?thx

Sure. Will take a look soon.

**Comment by kezhenxu94** [comment]
@CalvinKirs can we just cherry pick the related commits into the target branch?

**Comment by kezhenxu94** [comment]
Standalone server is mostly for convenient development, so I don't know how much it will benefit to cherry pick this to the 1.3.9 release, I think developers often develop on `dev` branch, right?

**Comment by CalvinKirs** [comment]
> @CalvinKirs can we just cherry pick the related commits into the target branch?

It’s very difficult, our branches are quite different :sob:

**Comment by CalvinKirs** [comment]
> Standalone server is mostly for convenient development, so I don't know how much it will benefit to cherry pick this to the 1.3.9 release, I think developers often develop on `dev` branch, right?

I totally agree with you. but the main purpose of doing this is to allow some new users to start easily and quickly. @dailidong

**Comment by davidzollo** [comment]
> > Standalone server is mostly for convenient development, so I don't know how much it will benefit to cherry pick this to the 1.3.9 release, I think developers often develop on `dev` branch, right?
> 
> I totally agree with you. but the main purpose of doing this is to allow some new users to start easily and quickly. @dailidong

Now there are many deployment steps for DolphinScheduler, especially for new users, what they want to feel most is to quickly experience whether this can meet his needs and how to quickly experience the functions of DolphinScheduler.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java [hunk 2]
```diff
@@ -56,6 +58,9 @@
     private int maxWaitTime;
 
     public String getServerList() {
+        if(StringUtils.isNotBlank(System.getProperty("zookeeper.quorum"))){
```
Comment (kezhenxu94) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java]: Seems it's not formatted

File: dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        new SpringApplicationBuilder(
+                ApiApplicationServer.class,
+                MasterServer.class,
+                WorkerServer.class
+        ).run(args);
+    }
+
+    private static void startAlertServer() {
+        AlertServer.getInstance().start();
+    }
```
Comment (kezhenxu94) [inline dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java:72]: Can you cherry-pick this part from the latest `dev` branch and uncomment line 61 to enable alert server?

```suggestion
    private static void startAlertServer() {
        final Path alertPluginPath = Paths.get(
                StandaloneServer.class.getProtectionDomain().getCodeSource().getLocation().getPath(),
                "../../../dolphinscheduler-alert-plugin/dolphinscheduler-alert-email/pom.xml"
        ).toAbsolutePath();
        if (Files.exists(alertPluginPath)) {
            System.setProperty("alert.plugin.binding", alertPluginPath.toString());
            System.setProperty("alert.plugin.dir", "");
        }
        AlertServer.getInstance().start();
    }
```

Comment (kezhenxu94) [reply dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java:72]: Well, the settings are totally different between this branch and `dev` branch?

I can see

```properties
alert.type=EMAIL
plugin.dir=/Users/xx/your/path/to/plugin/dir
```

vs 

```properties
alert.plugin.binding=
alert.plugin.dir=

**Review by kezhenxu94 — APPROVED** [review]
LGTM in terms of standalone server functionalities, I can see this error logs when running a simple shell task, might not be related to this but might be worthy to take a look.


```
2021-10-12 20:04:02.210 ERROR 76462 --- [ool-13-thread-2] o.a.d.remote.handler.NettyServerHandler  : process msg Command [type=TASK_EXECUTE_REQUEST, opaque=3, bodyLen=1510] error

java.lang.NullPointerException: null
	at org.apache.dolphinscheduler.server.worker.processor.TaskExecuteProcessor.getTaskLogPath(TaskExecuteProcessor.java:175) ~[classes/:na]
	at org.apache.dolphinscheduler.server.worker.processor.TaskExecuteProcessor.process(TaskExecuteProcessor.java:133) ~[classes/:na]
	at org.apache.dolphinscheduler.remote.handler.NettyServerHandler$1.run(NettyServerHandler.java:134) ~[classes/:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_292]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_292]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_292]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_292]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_292]

```

File: dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java (hunk 3 above)
Comment (CalvinKirs) [reply dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java:72]: yup, an Email Alert is that must be loaded. An email will be loaded automatically when AlertServer is started

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java (hunk 2 above)
Comment (CalvinKirs) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/zk/ZookeeperConfig.java]: Thx your remind, done

File: dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java (hunk 3 above)
Comment (CalvinKirs) [reply dolphinscheduler-standalone-server/src/main/java/org/apache/dolphinscheduler/server/StandaloneServer.java:72]: So we don’t need to set up email

**Review by kezhenxu94 — APPROVED** [review]
LGTM

**Review by lenboo — APPROVED** [review]
+1

**Pull request merged**